os/fs/smartfs/smartfs_smart.c

```diff
@@ -300,6 +303,23 @@ static int smartfs_open(FAR struct file *filep, const char *relpath, int oflags,
 	sf->curroffset = sizeof(struct smartfs_chain_header_s);
 	sf->currsector = sf->entry.firstsector;
 	sf->byteswritten = 0;
+	
+	if ((sf->bflags & SMARTFS_BFLAG_DIRTY) == 0) {
+		/* When using sector buffering, current sector with its header should
+		 * always be present in sf->buffer. Otherwise data corruption may arise
+		 * when writing.
+		 */
+
+		if (sf->currsector != SMARTFS_ERASEDSTATE_16BIT) {
+			smartfs_setbuffer(&readwrite, sf->currsector, 0, fs->fs_llformat.availbytes, (uint8_t *) sf->buffer);
```

Comment:
readwrite is defined under CONFIG_SMARTFS_USE_SECTOR_BUFFER
So this entire new code block should be too.